File: typescript/tests/client/fingerprint.test.ts
import { describe, expect, it } from 'bun:test';
import {
  collectFingerprint,
  defaultFingerprintSources,
  type FingerprintSource,
  fingerprintFromSources,
} from '../../src/client/fingerprint.ts';

describe('fingerprintFromSources', () => {
  it('returns a 64-char lowercase hex SHA-256', () => {
    const fp = fingerprintFromSources(['a', 'b']);
    expect(fp).toMatch(/^[0-9a-f]{64}$/);
  });

  it('is stable across input order', () => {
    expect(fingerprintFromSources(['a', 'b', 'c'])).toBe(fingerprintFromSources(['c', 'a', 'b']));
  });

  it('differs for different inputs', () => {
    expect(fingerprintFromSources(['a'])).not.toBe(fingerprintFromSources(['b']));
  });

  it('rejects empty source list', () => {
    expect(() => fingerprintFromSources([])).toThrow('at least one source');
  });

  it('rejects empty strings', () => {
    expect(() => fingerprintFromSources([''])).toThrow('non-empty');
  });

  it('rejects embedded newlines', () => {
    expect(() => fingerprintFromSources(['a\nb'])).toThrow('newline');
  });
});

describe('collectFingerprint', () => {
  const src = (name: string, value: string | null): FingerprintSource => ({
    name,
    collect: () => value,
  });

  it('filters out nulls but still hashes remaining', async () => {
    const fp = await collectFingerprint([src('a', 'hello'), src('b', null)]);
    expect(fp).toBe(fingerprintFromSources(['hello']));
  });

  it('throws when every source returns null', async () => {
    await expect(collectFingerprint([src('a', null), src('b', null)])).rejects.toThrow(
      'all configured fingerprint sources returned null',
    );
  });

  it('awaits async sources', async () => {
    const asyncSrc: FingerprintSource = {
      name: 'x',
      collect: () => Promise.resolve('async-value'),
    };
    const fp = await collectFingerprint([asyncSrc]);
    expect(fp).toBe(fingerprintFromSources(['async-value']));
  });

  it('empty strings from sources are treated as null', async () => {
    const fp = await collectFingerprint([src('a', 'real'), src('b', '')]);
    expect(fp).toBe(fingerprintFromSources(['real']));
  });
});

describe('defaultFingerprintSources', () => {
  it('requires a non-empty salt', () => {
    expect(() => defaultFingerprintSources('')).toThrow('appSalt must be a non-empty');
  });

  it('returns the four default sources in a fixed order', () => {
    const srcs = defaultFingerprintSources('my-app');
    expect(srcs.map((s) => s.name)).toEqual(['os.id', 'machine.id', 'net.primaryMac', 'app.salt']);
  });

  it('salt source collects deterministically', async () => {
    const srcs = defaultFingerprintSources('my-app');
    const salt = srcs.find((s) => s.name === 'app.salt');
    expect(salt).toBeDefined();
    if (!salt) return; // narrow for TS; unreachable after the expect above
    expect(await salt.collect()).toBe('salt:my-app');
  });

  it('fingerprint is stable across two collections on this machine', async () => {
    const srcs = defaultFingerprintSources('test-salt');
    const a = await collectFingerprint(srcs);
    const b = await collectFingerprint(srcs);
    expect(a).toBe(b);
  });

  it('custom source list replaces defaults entirely', async () => {
    const custom: readonly FingerprintSource[] = [
      { name: 'only-one', collect: () => 'custom-value' },
    ];
    const fp = await collectFingerprint(custom);
    expect(fp).toBe(fingerprintFromSources(['custom-value']));
  });
});

File: licensing/client/fingerprint_test.go
package client

import (
	"strings"
	"testing"
)

func TestFingerprintFromSources_Deterministic(t *testing.T) {
	sources := []string{"os:linux:amd64", "salt:app-v1", "machine:abc123"}
	h1, err := FingerprintFromSources(sources)
	if err != nil {
		t.Fatal(err)
	}
	// Different order, same result.
	reordered := []string{"salt:app-v1", "machine:abc123", "os:linux:amd64"}
	h2, err := FingerprintFromSources(reordered)
	if err != nil {
		t.Fatal(err)
	}
	if h1 != h2 {
		t.Fatalf("order should not affect hash: %s vs %s", h1, h2)
	}
	if len(h1) != 64 {
		t.Fatalf("expected 64-char hex, got %d", len(h1))
	}
}

func TestFingerprintFromSources_RejectsEmpty(t *testing.T) {
	if _, err := FingerprintFromSources(nil); err == nil {
		t.Fatal("expected error for empty source list")
	}
	if _, err := FingerprintFromSources([]string{""}); err == nil {
		t.Fatal("expected error for empty source string")
	}
	if _, err := FingerprintFromSources([]string{"a\nb"}); err == nil {
		t.Fatal("expected error for newline in source")
	}
}

func TestCollectFingerprint_RejectsAllEmpty(t *testing.T) {
	srcs := []FingerprintSource{
		emptySource{},
		emptySource{},
	}
	if _, err := CollectFingerprint(srcs); err == nil {
		t.Fatal("expected error when all sources return empty")
	}
}

func TestDefaultFingerprintSources_RequiresAppSalt(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected panic for empty appSalt")
		}
	}()
	DefaultFingerprintSources("")
}

func TestDefaultFingerprintSources_ProducesHash(t *testing.T) {
	srcs := DefaultFingerprintSources("test-salt-v1")
	h, err := CollectFingerprint(srcs)
	if err != nil {
		t.Fatalf("CollectFingerprint: %v", err)
	}
	if len(h) != 64 {
		t.Fatalf("expected 64-char hex, got %d: %s", len(h), h)
	}
	// Deterministic within a process.
	h2, err := CollectFingerprint(srcs)
	if err != nil {
		t.Fatal(err)
	}
	if h != h2 {
		t.Fatalf("non-deterministic: %s vs %s", h, h2)
	}
}

func TestSaltSource_Prefix(t *testing.T) {
	s := saltSource{salt: "foo"}
	v, err := s.Collect()
	if err != nil {
		t.Fatal(err)
	}
	if !strings.HasPrefix(v, "salt:") {
		t.Fatalf("expected salt: prefix, got %q", v)
	}
}

type emptySource struct{}

func (emptySource) Name() string             { return "empty" }
func (emptySource) Collect() (string, error) { return "", nil }
